### DISCUSSION TITLE
Increasing speed of algorithms using cython

### DISCUSSION THREAD

**Discussion by matusvalo** [networkx/networkx#5851] [Ideas]
Hi all,

I have played again (my previous attempt is #4130) speeding up networkx using cython. Currently, I have focused on following algorithms.
### Test:
```python
path_len = [1, 10, 20, 50, 100, 200, 500, 1000, 2000, 5000, 10000, 50000]
times_patched = []
for i in path_len:
    times_patched.append(
        timeit.timeit(stmt='list(nx.strongly_connected_components(G))', setup=f'import networkx as nx; G = nx.path_graph({i}, create_using=nx.DiGraph)', number=100)
    )
```
### Result:
![image](https://user-images.githubusercontent.com/827060/177838934-abb204de-e986-470d-a28c-0b82681ad5ce.png)
### Implementation: 
Code was modified only by annotating variable `i` as `cython.int`:
```diff
@@ -74,7 +75,7 @@ def strongly_connected_components(G):
     lowlink = {}
     scc_found = set()
     scc_queue = []
-    i = 0  # Preorder counter
+    i: cython.int = 0  # Preorder counter
     neighbors = {v: iter(G[v]) for v in G}
     for source in G:
         if source not in scc_found:
@@ -95,9 +96,9 @@ def strongly_connected_components(G):
                     for w in G[v]:
                         if w not in scc_found:
                             if preorder[w] > preorder[v]:
-                                lowlink[v] = min([lowlink[v], lowlink[w]])
+                                lowlink[v] = min(lowlink[v], lowlink[w])
                             else:
-                                lowlink[v] = min([lowlink[v], preorder[w]])
+                                lowlink[v] = min(lowlink[v], preorder[w])
                     queue.pop()
```

### Test:
```python
path_len = [1, 10, 20, 50, 100, 200, 500, 1000, 2000, 5000, 10000, 50000]
times_patched = []
for i in path_len:
    times_patched.append(
        timeit.timeit(stmt='list(nx.biconnected_components(G))', setup=f'import networkx as nx; G = nx.circular_ladder_graph({i})', number=100)
    )
```
### Result:
![image](https://user-images.githubusercontent.com/827060/177840301-71e8a896-39e5-43c1-935e-8759898b5368.png)
### Implementation:
Code was again just slightly annotated:
```diff
@@ -330,7 +331,7 @@ def articulation_points(G):
 @not_implemented_for("directed")
-def _biconnected_dfs(G, components=True):
+def _biconnected_dfs(G, components: cython.bint=True):
     # depth-first search algorithm to generate articulation points
     # and biconnected components
     visited = set()
@@ -339,7 +340,7 @@ def _biconnected_dfs(G, components=True):
             continue
         discovery = {start: 0}  # time of first discovery of node during search
         low = {start: 0}
-        root_children = 0
+        root_children: cython.int = 0
         visited.add(start)
```

### Test
```python
path_len = [1, 10, 20, 50, 100, 200, 500, 1000, 2000, 5000, 10000, 50000]
times_patched = []
for i in path_len:
    times_patched.append(
        timeit.timeit(stmt='list(nx.weakly_connected_components(G))', setup=f'import networkx as nx; G = nx.DiGraph(); G.add_nodes_from(range({i}))', number=100)
    )
```
### Result:
![image](https://user-images.githubusercontent.com/827060/177841182-a29ad61d-a97d-4334-adfe-ec4be26ec67a.png)
### Implementation
No changes in code
```

## `nx.eulerian_circuit()`
### Test
```python
path_len = [1, 11, 21, 51, 101, 201, 301, 501, 701, 1001, 2001, 3001]
times_patched = []
for i in path_len:
    times_patched.append(timeit.timeit(stmt='list(nx.eulerian_circuit(G))', setup=f'import networkx as nx; G = nx.complete_graph({i})', number=10))
```
### Result:
![image](https://user-images.githubusercontent.com/827060/177841532-e9267af0-f0dd-43ea-b034-e336cbc02fb7.png)

### Implementation
`nx.eulerian_curcuit()` did not show significant improvements by naive annotation and compilation. The speedup was achieved by:
1. annotation of simple variables
2. Creation fastpaths for following methods:
    a. `Graph.degre()`, `Graph.remove_edge()`, `Graph.edges()`
    b. `DiGraph.out_degree()`, `DiGraph.in_degree()`, `DiGraph.out_edges()`, `DiGraph.remove_edge()`
This approach was already proposed in issue #4895, this approach further improves it using Cython syntax/optimizations.

> Note: I will gladly provide used code PoC when there will be interest.

## Discussion

Cython is already well known and widely used in scientific community and beyond:
* https://github.com/scipy/scipy/search?l=cython
* https://github.com/aio-libs/aiohttp/search?l=cython
* https://github.com/pandas-dev/pandas/search?l=cython
...

The benefit of cython is that user has full control over optimisation. On the other hand this is drawback because the cython adds  some complexity to the code. Other benefit of this approach is that the code base can be optimised incrementally. Mypyc [1] provides transparent optimisation of code but requires fully annotated code (plus compiling of code by mypyc has some side effects like freezing classes which can yield significant code modification needed for adoption).

## Proposal

Incremental adoption of cython algorithm per algorithm:
1. compile during build some handpicked algorithms. Pure python mode [2] can be used so networkx can be still used without need of compilation (this will keep low adoption barrier for networkx).
2. create fastpaths for several Graph methods and compile them by cython. Again pure python mode can be used. There is already issue for that #4895
3. Use more aggressive (low level) optimizations. Several data structures can be rewritten in Cython to more low level implementation. E.g. Using  C++ stdlib containers store C types instead of python objects [3]

> Note: Of course I am not proposing implementing all points 1. - 3. My point is that at least point 1. can be implemented since it is basically for free.   

[1] https://github.com/python/mypy/tree/master/mypyc
[2] https://cython.readthedocs.io/en/latest/src/tutorial/pure.html
[3] https://cython.readthedocs.io/en/latest/src/userguide/wrapping_CPlusPlus.html

**Comment by rossbar** [comment]
I don't think the outlook has changed re: making NetworkX no longer a pure-Python package.

If you're looking to contribute cythonized graph algorithms you might want to check out the `scipy.sparse.csgraph` package, which has cythonized implementations of some common algorithms.
